### PR TITLE
[Syntax] Parse invalid characters as trivia

### PR DESCRIPTION
**Pull request by omochi**
This PR update Lexer to parse invalid characters as trivia.

Invalid chars means here is invalid UTF-8 byte sequence and unicode code point which is invalid in Swift source.
They are just skipped in lexImpl before.

By this PR, libSyntax get perfect round trip with source code which has arbitrary binary even if it contains broken UTF-8 byte sequence.

This is my third retry for same purpose.
First is https://github.com/apple/swift/pull/14967.
Second is https://github.com/apple/swift/pull/14979.
In this time, I covered all considerations and retrieved review.

I rearranged concepts around here.
There are 2 kind of things.

First is `unknown token`.
There are some rules to lex `unknown token` in Lexer.
When hit these tokens in `lexImpl`, Lexer need to stop reading and position start of token.

Second is `garbage text trivia`.
They come from invalid byte sequences and unicode codepoint which is invalid in Swift source.

To detect both cases in `lexTrivia`,
I added some new `case` branches and I minimize them.
Specifically, to minimize,
`case` below are handled in `advanceIfValidStartOfIdentifier` in `default` branch.

```
case 'A': case 'B': case 'C': case 'D': case 'E': case 'F': case 'G':
case 'H': case 'I': case 'J': case 'K': case 'L': case 'M': case 'N':
case 'O': case 'P': case 'Q': case 'R': case 'S': case 'T': case 'U':
case 'V': case 'W': case 'X': case 'Y': case 'Z':
case 'a': case 'b': case 'c': case 'd': case 'e': case 'f': case 'g':
case 'h': case 'i': case 'j': case 'k': case 'l': case 'm': case 'n':
case 'o': case 'p': case 'q': case 'r': case 's': case 't': case 'u':
case 'v': case 'w': case 'x': case 'y': case 'z':
case '_': 
```

`case` below are handled in `advanceIfValidStartOfOperator` in `default` branch too.

```
case '%': case '!': case '?':case '=':
case '-': case '+': case '*': 
case '&': case '|': case '^': case '~': case '.':
```

First 4 commits does not change behavior of Lexer.
They are step by step refactoring to prepare new function.

Last commit implement lexing and adding test cases.

Please review this. @rintaro @harlanhaskins @nkcsgexi

### CODE REVIEW

**Comment by rintaro** [comment]
> I added some new case branches and I minimize them.

For performance reasons, I think we should handle both `isValidIdentifierStartCodePoint` and `isOperatorStartCodePoint` cases in `switch` in `lexTrivia`.
Since this switch is super-hot path, we want to avoid performance regressions at all costs.

File: test/Syntax/tokens_unknown_and_invalid.swift [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+// [0xE1, 0x9A, 0x80] in UTF-8 is U+1680.
+// This character is invalid for swift source.
+// It becomes garbage trivia.
+// Marker(5) is replaced to this sequence.
+
+// RUN: -e 's/'$(echo -ne "\x5a5")'/'$(echo -ne "\xe1\x9a\x80")'/g' \
+
+// RUN: > %t
+
+// RUN: %swift-syntax-test -input-source-filename %t -dump-full-tokens 2>&1 | %FileCheck %t
```
Comment (rintaro) [inline test/Syntax/tokens_unknown_and_invalid.swift]: Could you consolidate `round_trip_unknown_and_invalid.swift` into this file? i.e.
```swift
// RUN: %swift-syntax-test -input-source-filename %t -dump-full-tokens 2>&1 | %FileCheck %t
// RUN: %round-trip-syntax-test --swift-syntax-test %swift-syntax-test --file %t
```

Comment (omochi) [reply test/Syntax/tokens_unknown_and_invalid.swift]: OK!

**Comment by omochi** [comment]
> Since this switch is super-hot path, we want to avoid performance regressions (especially in normal scenario) at all costs.
Ok, I expanded these character cases.

Updated.

**Comment by rintaro** [comment]
Thanks!

@swift-ci Please smoke test

**Comment by rintaro** [comment]
@nkcsgexi Please merge this if you are OK.

**Comment by nkcsgexi** [comment]
LGTM! 🚢

**Comment by omochi** [comment]
Thank you all 😄

**Pull request merged**